packages/graalvm/src/main/kotlin/elide/runtime/feature/engine/AbstractStaticNativeLibraryFeature.kt

```diff
@@ -13,26 +13,116 @@
+import com.oracle.svm.hosted.jni.JNIFeature
+import org.graalvm.nativeimage.hosted.Feature
 import org.graalvm.nativeimage.hosted.Feature.BeforeAnalysisAccess
+import java.io.FileNotFoundException
+import java.io.InputStream
+import java.lang.reflect.Field
+import java.lang.reflect.Method
+import java.nio.file.FileSystems
+import java.nio.file.Files
+import java.nio.file.Path
+import kotlin.io.path.Path
 import elide.runtime.feature.NativeLibraryFeature
 import elide.runtime.feature.NativeLibraryFeature.NativeLibInfo
+import elide.runtime.feature.NativeLibraryFeature.NativeLibType
 import elide.runtime.feature.NativeLibraryFeature.NativeLibType.SHARED
 import elide.runtime.feature.NativeLibraryFeature.NativeLibType.STATIC
 
-/**
- * TBD.
- */
-internal abstract class AbstractStaticNativeLibraryFeature : NativeLibraryFeature {
-  /**
-   * TBD.
-   */
-  protected fun libraryNamed(name: String): NativeLibInfo = NativeLibInfo.of(name)
-
-  /**
-   * TBD.
-   */
+public abstract class AbstractStaticNativeLibraryFeature : NativeLibraryFeature {
+  protected fun libraryNamed(
+    name: String,
+    vararg layout: String,
+    type: NativeLibType = STATIC,
+    linkName: String = name,
+    builtin: Boolean = true,
+    registerJni: Boolean = true,
+    registerPrefix: Boolean = builtin,
+    eager: Boolean = builtin,
+    absolutePath: Path? = null,
+    initializer: Boolean = false,
+  ): NativeLibInfo = NativeLibInfo.of(
+    name,
+    *layout,
+    type = type,
+    prefix = registerPrefix,
+    jni = registerJni,
+    builtin = builtin,
+    eager = eager,
+    linkName = linkName,
+    absolutePath = absolutePath,
+    initializer = initializer,
+  )
+
+  // Scan each provided JAR file system path (as a ZipFileSystem) for the resource at the provided path; when a match is
+  // found, it is returned as an `InputStream`, otherwise an error is thrown if the resource is not found.
+  private fun scanJarsForResource(jars: List<Path>, path: String): InputStream {
+    for (jar in jars) {
+      val env = mapOf("enablePosixFileAttributes" to "true")
+      requireNotNull(FileSystems.newFileSystem(jar, env)) { "JAR at path not found: '$jar'" }.use {
+        try {
+          return it.getPath(path).toUri().toURL().openStream()
+        } catch (fnf: FileNotFoundException) {
+          // ignore
+        }
+      }
+    }
+    error("Resource not found in any JAR: '$path' in JARs: '${jars.joinToString(", ")}'")
+  }
+
+  private fun BeforeAnalysisAccess.scanForResource(jarToken: String, path: String): InputStream? {
+    val candidates = applicationClassPath.filter { it.toString().contains(jarToken) }
+    if (candidates.isEmpty()) {
+      return null
+    }
+    return scanJarsForResource(candidates, path)
+  }
+
+  protected fun BeforeAnalysisAccess.unpackLibrary(
+    jar: String,
+    name: String,
+    arch: String,
+    vararg path: String,
+    renameTo: ((String) -> String)? = null,
+    cbk: (() -> Unit)? = null,
+  ) {
```

Comment:
one can now "unpack libraries" into the Elide natives root at image build time; these libraries (expected to be `.so`, `.dylib`, `.a`, and so on), are eligible for static JNI.